File: src/main/kotlin/net/echo/little_johns/block/custom/EcoFriendlyWoodVeneersBlock.java
package net.echo.little_johns.block.custom;


import com.mojang.serialization.MapCodec;
import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.block.Waterloggable;
import net.minecraft.fluid.FluidState;
import net.minecraft.fluid.Fluids;
import net.minecraft.item.ItemPlacementContext;
import net.minecraft.state.StateManager;
import net.minecraft.state.property.BooleanProperty;
import net.minecraft.state.property.Properties;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.util.math.random.Random;
import net.minecraft.world.WorldView;
import net.minecraft.world.tick.ScheduledTickView;

public class EcoFriendlyWoodVeneersBlock
        extends MultifaceBlock
        implements Waterloggable {

    public static final MapCodec<EcoFriendlyWoodVeneersBlock> CODEC = EcoFriendlyWoodVeneersBlock.createCodec(EcoFriendlyWoodVeneersBlock::new);
    private static final BooleanProperty WATERLOGGED = Properties.WATERLOGGED;

    public MapCodec<EcoFriendlyWoodVeneersBlock> getCodec() {
        return CODEC;
    }

    public EcoFriendlyWoodVeneersBlock(Settings settings) {
        super(settings);
        this.setDefaultState((BlockState)this.getDefaultState().with(WATERLOGGED, false));
    }

    @Override
    protected void appendProperties(StateManager.Builder<Block, BlockState> builder) {
        super.appendProperties(builder);
        builder.add(WATERLOGGED);
    }

    protected BlockState getStateForNeighborUpdate(BlockState state, WorldView world, ScheduledTickView tickView, BlockPos pos, Direction direction, BlockPos neighborPos, BlockState neighborState, Random random) {
        if ((Boolean)state.get(WATERLOGGED)) {
            tickView.scheduleFluidTick(pos, Fluids.WATER, Fluids.WATER.getTickRate(world));
        }

        return super.getStateForNeighborUpdate(state, world, tickView, pos, direction, neighborPos, neighborState, random);
    }


    @Override
    protected boolean canReplace(BlockState state, ItemPlacementContext context) {
        return context.getStack().isOf(this.asItem());
    }



    @Override
    protected FluidState getFluidState(BlockState state) {
        if (state.get(WATERLOGGED).booleanValue()) {
            return Fluids.WATER.getStill(false);
        }
        return super.getFluidState(state);
    }


}

File: src/main/kotlin/net/echo/little_johns/block/custom/EcoFriendlyWoodVeneersBlock.kt
package net.echo.little_johns.block.custom

import com.mojang.serialization.MapCodec
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.item.ItemPlacementContext
import net.minecraft.state.StateManager


class EcoFriendlyWoodVeneersBlock(settings: Settings) : ModdedMultifaceBlock(settings) {

    override fun getCodec(): MapCodec<EcoFriendlyWoodVeneersBlock> = CODEC

    override fun canReplace(state: BlockState, context: ItemPlacementContext): Boolean {
        return context.stack.isOf(this.asItem())
    }

    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) {
        super.appendProperties(builder)
    }

    init {
        defaultState = defaultState.with(WATERLOGGED, false)
    }

    companion object {
        val CODEC: MapCodec<EcoFriendlyWoodVeneersBlock> = createCodec(::EcoFriendlyWoodVeneersBlock)
    }
}
